Add new endpoint for joining a cluster

mbarnes opened the pull request:
This was an idea I had to help with the host auto-join feature.

Programmatically determining a host's own IP address to send to the Commissaire server can be non-trivial given the possibility of multiple network devices.  And even if you know the right device name, parsing it out from `ip address` or `ifconfig` is still a pain.

This helps alleviate the problem by offering a convenient endpoint for hosts to add themselves directly to a cluster.

***Update:*** *After further discussion the endpoint was changed to `/api/v0/host`.*

**Endpoint:** `/api/v0/cluster/{NAME}/join`

Accepts a `PUT` request from the host wishing to join cluster `{NAME}`.
    
This is equivalent to:
```    
   PUT /api/v0/host/{ADDRESS}
    
   {
      cluster: {NAME}
      ...
   }
```
except the host's `{ADDRESS}` is derived from the request itself.

mbarnes commented:
@cgwalters-bot r+ e5edbdb

ashcrow commented:
The idea of using the requesting address makes a lot of sense. Though why not update the current host endpoint instead of creating a new one?

ashcrow commented:
@cgwalters-bot try

cgwalters commented:
@cgwalters-bot retry

ashcrow commented:
@cgwalters beat me to it, just read I should have used retry :smile:

ashcrow commented:
@cgwalters-bot r+ da7a878ce23bc33b290c45f7f5cbf003617a0f7f

ashcrow commented:
@cgwalters-bot r- da7a878ce23bc33b290c45f7f5cbf003617a0f7f
@cgwalters-bot r+ da7a878ce23bc33b290c45f7f5cbf003617a0f7f

ashcrow commented:
Messing with the bot a bit here to see what happens :smile: 

@cgwalters-bot clean

mbarnes commented:
> The idea of using the requesting address makes a lot of sense. Though why not update the current host endpoint instead of creating a new one?

Maybe I misunderstand what you mean, but the host endpoint URL embeds the host's IP address, which defeats the purpose here.  I'm trying to avoid forcing the host to know its own IP address before automatically registering itself to Commissaire.

Using the endpoint of the cluster it wants to join seemed the most straight-forward.

I'm open to other ideas though.

ashcrow commented:
@mbarnes I see what you mean. Since the ```GET``` will always required a name it would be awkward. 

What about having a ```PUT```/```POST``` ```/api/v0/host/``` endpoint that takes the current data ```/api/v0/host/{IP}``` takes sans the ```IP``` in the URL. The resulting instance would be available at ```/api/v0/host/{IP}``` and we can change ```/api/v0/host/{IP}``` to be ```GET```/``DELETE``` only.

ashcrow commented:
@cgwalters-bot clean

cgwalters commented:
Hmm...I'd have expected "clean" to work.  Let me debug.

cgwalters commented:
Oh although, I bet this needs to be rebased on master so we have the webhook in the PR series.

mbarnes commented:
@ashcrow: I was with you until that last bit:
> and we can change `/api/v0/host/{IP}` to be `GET`/`DELETE` only

Wouldn't that impose a new requirement that a host creation request has to come from the host itself?  Unless that was always the intention, I think we'd want to keep both ways for the sake of administrators.

If you still prefer a `PUT` on `/api/v0/host` I can rework it.

ashcrow commented:
@mbarnes good point. Scratch the last bit. Yeah throw it on ```/api/v0/host/``` and I think this will be nice and clean.

ashcrow commented:
@mbarnes mind rebasing to see if that makes homu happier?

cgwalters commented:
@ashcrow You'll need to re-`r+` with new commit ID of `d3fa8b5`

ashcrow commented:
@cgwalters-bot r+ d3fa8b513afb0dae99a1f098652b8656ab86f37d

mbarnes commented:
Why was this PR closed?

cgwalters commented:
It was closed because it was merged...sorry I know it's a bit confusing right now.  See https://github.com/barosl/homu/pull/139

mbarnes commented:
This accidentally got merged to `master` prematurely, so here's what I did for the rework.
 
* Left the 1st commit (https://github.com/projectatomic/commissaire/commit/4969962735b7c31432a372f5ee0f6afd1651468d) merged since it was just a util function and didn't need changed.  If there's corrections to be made there I'll do so in a new commit.
* Reverted the 2nd commit on `master` and rebased this branch on it.
* Added :arrow_up: for the `/api/v0/host` endpoint as discussed.

For the endpoint docs I just added a note to the host creation part.  This is kind of a special case so it didn't seem worth a whole new section to me.

ashcrow commented:
@cgwalters-bot try

cgwalters commented:
@ashcrow I think that may currently not do anything because the PR already passed?  Ah yes, look at:

```
def process_queue(...):
    for repo_label, repo in repos.items():
        for state in repo_states:
            if state.status == '' and state.try_:
                if start_build(state, repo_cfgs, buildbot_slots, logger, db, git_cfg):
                    return
```

Since `status != ''` we don't try the merge build.  I'd say that's a conceptual bug - Homu should support it.

ashcrow commented:
@cgwalters ah ok.

@cgwalters-bot  r+ 71d90496cdf8a7b958ec8001e0119c90925a71bb
